File: 4.8-bc/py/stage-2-feature-first/app/features/billing/handler.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.features.billing.repository import Repository
from app.features.billing.service import InvalidPlanError, Service
from app.shared.db import get_session

router = APIRouter(tags=["billing"])


class SubscribeReq(BaseModel):
    user_id: str
    plan: str


class SubscribeResp(BaseModel):
    subscription_id: str
    plan: str
    next_charge_at: datetime


@router.post(
    "/subscriptions",
    status_code=status.HTTP_201_CREATED,
    response_model=SubscribeResp,
)
async def subscribe(
    req: SubscribeReq, session: AsyncSession = Depends(get_session)
) -> SubscribeResp:
    if not req.plan:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="invalid_request"
        )
    try:
        uid = UUID(req.user_id)
    except (ValueError, TypeError) as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="invalid_user_id"
        ) from exc
    svc = Service(Repository(session))
    try:
        sub = await svc.subscribe(uid, req.plan)
    except InvalidPlanError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"error": "billing.invalid_plan", "message": "plan must be basic|pro|enterprise"},
        ) from exc
    return SubscribeResp(
        subscription_id=str(sub.id),
        plan=sub.plan,
        next_charge_at=sub.next_charge_at,
    )

File: 4.8-bc/ts/stage-2-feature-first/src/features/billing/handler.ts
import type { FastifyInstance } from "fastify";

import { writeError } from "../../shared/httputil.js";
import type { Service } from "./service.js";

interface SubscribeBody {
  user_id?: string;
  plan?: string;
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.post<{ Body: SubscribeBody }>("/subscriptions", async (req, reply) => {
      const body = req.body ?? {};
      if (!body.plan) {
        return reply.code(400).send({ error: "invalid_request" });
      }
      if (!body.user_id || !UUID_RE.test(body.user_id)) {
        return reply.code(400).send({ error: "invalid_user_id" });
      }
      try {
        const sub = await this.svc.subscribe(body.user_id, body.plan);
        return reply.code(201).send({
          subscription_id: sub.id,
          plan: sub.plan,
          next_charge_at: sub.nextChargeAt,
        });
      } catch (err) {
        req.log.error({ err }, "subscribe failed");
        return writeError(reply, err);
      }
    });
  }
}
